### PR TITLE
User can login and log out

### PR DESCRIPTION
**Pull request by GergKllai1**
# PT Link

[User can login and log out](https://www.pivotaltracker.com/story/show/163530608)

# User Story

```
As a user
In order to access the features on the site
I would like to be able to log in and log out
```

# Screenshots

feb-01-2019 10-33-57

# Description of PR

- Adds login form component
- Sends axios post request to sessions controller
- Tailors session controller to accept incoming data and send back error response
- Conditions sessions and authentication links based on whether user is logged in or not
- Adds basic styling
- Refactors components into matching folders
- Refactors feature tests

### CODE REVIEW

File: app/controllers/devise/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      else
+        render json: {errors: "Invalid Email or password."}
+      end
+    end
+  
+    def destroy
+      signed_out = (Devise.sign_out_all_scopes ? sign_out : sign_out(resource_name))
+      set_flash_message! :notice, :signed_out if signed_out
+      yield if block_given?
+      # respond_to_on_destroy
```
Comment (HannaSwDn) [inline app/controllers/devise/sessions_controller.rb]: could you remove this commented out line? :)

File: client/bundles/Main/components/Main.jsx [hunk 2]
```diff
@@ -1,66 +1,57 @@
 import React from "react";
-import AllSessions from "./AllSessions";
-import { NavLink, Link } from 'react-router-dom'
+import MainSessionsDisplay from "./sessions/MainSessionsDisplay";
+import MainNavLinks from "./MainNavLinks";
+
+
```
Comment (schneiderman805) [inline client/bundles/Main/components/Main.jsx]: please remove empty line

**Review by schneiderman805 — CHANGES_REQUESTED** [review]

File: app/controllers/devise_controller.rb [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                   resource_class resource_params devise_mapping)
+      helper_method(*helpers)
+    end
+  
+    prepend_before_action :assert_is_devise_resource!
+    respond_to :html if mimes_for_respond_to.empty?
+  
+    # Override prefixes to consider the scoped view.
+    # Notice we need to check for the request due to a bug in
+    # Action Controller tests that forces _prefixes to be
```
Comment (HannaSwDn) [inline app/controllers/devise_controller.rb]: could you remove these comments?

File: app/controllers/devise_controller.rb [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  
+    # Attempt to find the mapped route for devise based on request path
+    def devise_mapping
+      @devise_mapping ||= request.env["devise.mapping"]
+    end
+  
+    # Checks whether it's a devise mapped resource or not.
+    def assert_is_devise_resource! #:nodoc:
+      unknown_action! <<-MESSAGE unless devise_mapping
+  Could not find devise mapping for path #{request.fullpath.inspect}.
```
Comment (HannaSwDn) [inline app/controllers/devise_controller.rb]: Could you indent this part of the file

File: app/controllers/devise_controller.rb [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  
+      if notice
+        set_flash_message! :notice, notice
+        true
+      end
+    end
+  
+    # Sets the flash message with :key, using I18n. By default you are able
+    # to set up your messages using specific resource scope, and if no message is
+    # found we look to the default scope. Set the "now" options key to a true
```
Comment (HannaSwDn) [inline app/controllers/devise_controller.rb]: do we need these comments?

File: app/controllers/devise_controller.rb [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    # Get message for given
+    def find_message(kind, options = {})
+      options[:scope] ||= translation_scope
+      options[:default] = Array(options[:default]).unshift(kind.to_sym)
+      options[:resource_name] = resource_name
+      options = devise_i18n_options(options)
+      I18n.t("#{options[:resource_name]}.#{kind}", options)
+    end
+  
+    # Controllers inheriting DeviseController are advised to override this
```
Comment (HannaSwDn) [inline app/controllers/devise_controller.rb]: do we need these comments?

File: app/controllers/devise_controller.rb (hunk 5 above)
Comment (HannaSwDn) [reply app/controllers/devise_controller.rb]: did they get scaffolded with the file?

File: client/bundles/Main/components/Main.jsx (hunk 2 above)
Comment (kianaditya) [reply client/bundles/Main/components/Main.jsx]: Done

**Review by johannaanden — APPROVED** [review]
LGTM

File: app/controllers/devise/sessions_controller.rb (hunk 1 above)
Comment (GergKllai1) [reply app/controllers/devise/sessions_controller.rb]: Done

File: app/controllers/devise_controller.rb (hunk 3 above)
Comment (GergKllai1) [reply app/controllers/devise_controller.rb]: Done

File: app/controllers/devise_controller.rb (hunk 4 above)
Comment (GergKllai1) [reply app/controllers/devise_controller.rb]: Removed file

File: app/controllers/devise_controller.rb (hunk 5 above)
Comment (GergKllai1) [reply app/controllers/devise_controller.rb]: Removed the file

File: app/controllers/devise_controller.rb (hunk 6 above)
Comment (GergKllai1) [reply app/controllers/devise_controller.rb]: Removed the file

**Review by schneiderman805 — APPROVED** [review]

**Review by magnus-thor — APPROVED** [review]
Good job

**Pull request merged**